### PR TITLE
Check the existence of bcls

### LINKED ISSUE #382: Searching for current cycle when they do not exist yet breaks
We should check that the cycles exist then take the last one or return 0
see:
https://github.com/EdinburghGenomics/Analysis-Driver/blob/096af2f460317f81de749c5de256b8e338a9b83b/analysis_driver/pipelines/demultiplexing.py#L241

### PR DESCRIPTION
**Pull request by tcezard**
Check that the bcl exists before the returning the number of existing cycles
closes #382

### CODE REVIEW

File: analysis_driver/quality_control/interop_metrics.py [hunk 1]
```diff
@@ -109,9 +111,22 @@ def detect_bad_cycles(self):
         return bad_cycle_per_lanes
 
 
-def get_cycles_extracted(run_dir):
+def get_last_cycles_with_existing_bcls(run_dir):
+    """
+    This function check the extracted cycle from the interop and confirm the presence of the bcl files o the filesystem.
```
Comment (lukbut) [inline analysis_driver/quality_control/interop_metrics.py]:
```suggestion
    This function checks the extracted cycle from the interop and confirm the presence of the bcl files on the filesystem.
```

File: analysis_driver/quality_control/interop_metrics.py [hunk 2]
```diff
@@ -109,9 +111,22 @@ def detect_bad_cycles(self):
...
+    the confirmation is only performed from the last cycles and the first full confirmed cycle is returned.
```
Comment (lukbut) [inline analysis_driver/quality_control/interop_metrics.py]:
```suggestion
    The confirmation is only performed from the last cycles and the first full confirmed cycle is returned.
```

File: analysis_driver/quality_control/interop_metrics.py [hunk 3]
```diff
@@ -109,9 +111,22 @@ def detect_bad_cycles(self):
...
+    :returns: the last cycle of the run with existing bcls
+    """
     run_metrics = RunMetrics()
     run_metrics.read(run_dir)
     extraction_metrics = run_metrics.extraction_metric_set()
-    return sorted(extraction_metrics.cycles())
+    all_cycles = extraction_metrics.cycles()
+    last_complete_cycles = 0
+    for cycle in all_cycles[::-1]:
+        all_bcls = glob.glob(os.path.join(run_dir, 'Data/Intensities/BaseCalls/L00*/C%s.1/*.bcl.gz' % cycle))
```
Comment (lukbut) [inline analysis_driver/quality_control/interop_metrics.py]: Just to confirm, you are checking the number of BCLs in each cycle going backwards from the last cycle, and comparing it with the one you are extracting from the metrics, and then returning the last completed cycle (which is the one whose number of BCLs match the number extracted from the metrics.

File: tests/test_pipelines/test_demultiplexing.py [hunk 4]
```diff
@@ -107,8 +107,8 @@ def test_run(self):
                             lims_run=Mock(udf={'Run Status': 'RunStarted'}))
 
         # cycle extracted states 310 cycles done
-        pcycles = patch('analysis_driver.quality_control.interop_metrics.get_cycles_extracted',
-                        return_value=range(1, 311))
+        pcycles = patch('analysis_driver.quality_control.interop_metrics.get_last_cycles_with_existing_bcls',
+                        return_value=310)
```
Comment (lukbut) [inline tests/test_pipelines/test_demultiplexing.py:111]: Why 310 not 311? Is it because you are now compensating in the function?

**Review by lukbut — COMMENTED** [review]
I have gone through the code and have the following comments, but I do not feel I can vouch for the changes in technical functionality as I am not familiar with this part of the code at all.

File: tests/test_pipelines/test_demultiplexing.py (hunk 4 above)
Comment (tcezard) [reply tests/test_pipelines/test_demultiplexing.py:111]: The previous function returned a list of cycle which is mimicked using `range`. in `range` the end is not inclusive.
The new function returns only the last cycle value.

File: analysis_driver/quality_control/interop_metrics.py (hunk 3 above)
Comment (tcezard) [reply analysis_driver/quality_control/interop_metrics.py]: That's right. I've added a comment to clarify

File: analysis_driver/quality_control/interop_metrics.py [hunk 5]
```diff
@@ -109,9 +111,22 @@ def detect_bad_cycles(self):
...
     run_metrics = RunMetrics()
     run_metrics.read(run_dir)
     extraction_metrics = run_metrics.extraction_metric_set()
-    return sorted(extraction_metrics.cycles())
-
+    all_cycles = sorted(extraction_metrics.cycles())
+    last_complete_cycles = 0
+    # start from the last cycle and walk back until found a cycle with all the bcl present.
+    for cycle in all_cycles[::-1]:
+        all_bcls = glob.glob(os.path.join(run_dir, 'Data/Intensities/BaseCalls/L00*/C%s.1/*.bcl.gz' % cycle))
```
Comment (mwhamgenomics) [inline analysis_driver/quality_control/interop_metrics.py]: Ideally we'd want to be consistent with the rest of the repo and use `egcg_core.util.find_files`:

```python
util.find_files(run_dir, 'Data', 'Intensities', 'BaseCalls', 'L00*', 'C%s.1' % cycle, '*.bcl.gz')
```

But this should be okay until we port the pipeline to MacOS 9 😐.

**Review by mwhamgenomics — APPROVED** [review]

File: analysis_driver/quality_control/interop_metrics.py (hunk 5 above)
Comment (tcezard) [reply analysis_driver/quality_control/interop_metrics.py]: I've changed it to `util.find_files(run_dir, ...`

**Comment by tcezard** [comment]
This PR will not go through validation as it address an urgent bug. The change does not modify any feature that can influence the data or the metrics but only when the data will start being processed.

**Pull request merged**